lisa/tools/fio.py

```diff
@@ -0,0 +1,202 @@
+
+    def _get_iops_latency_by_qdepth(
+        self, qdepth: int, fio_results_list: List[FIOResult]
+    ) -> Dict[str, Decimal]:
+        mode_iops_latecny: Dict[str, Any] = {}
+        for fio_result in fio_results_list:
+            if qdepth == fio_result.qdepth:
+                mode_iops_latecny[f"{fio_result.mode}_iops"] = fio_result.iops
+                mode_iops_latecny[f"{fio_result.mode}_lat_usec"] = fio_result.latency
+                mode_iops_latecny["iodepth"] = fio_result.iodepth
+                mode_iops_latecny["qdepth"] = fio_result.qdepth
+        return mode_iops_latecny
+
+    def install(self) -> bool:
+        posix_os: Posix = cast(Posix, self.node.os)
+        posix_os.install_packages("fio")
+        if not self._check_exists():
+            self._install_from_src()
+        return self._check_exists()
+
+    def launch(
+        self,
+        name: str,
+        filename: str,
+        mode: str,
+        iodepth: int,
+        numjob: int,
+        time: int = 60,
+        block_size: str = "4K",
+        size_gb: int = 0,
+        direct: bool = True,
+        gtod_reduce: bool = True,
+        ioengine: str = "libaio",
+        group_reporting: bool = True,
+        overwrite: bool = False,
+    ) -> FIOResult:
+        cmd = (
+            f"--ioengine={ioengine} --bs={block_size} --filename={filename} "
+            f"--readwrite={mode} --runtime={time} --iodepth={iodepth} "
+            f"--numjob={numjob} --name={name}"
+        )
+        if direct:
+            cmd += " --direct=1"
+        if gtod_reduce:
+            cmd += " --gtod_reduce=1"
+        if size_gb:
+            cmd += f" --size={size_gb}G"
+        if group_reporting:
+            cmd += " --group_reporting"
+        if overwrite:
+            cmd += " --overwrite=1"
+        result = self.run(
+            cmd,
+            force_run=True,
+            sudo=True,
+            expected_exit_code=0,
+            expected_exit_code_failure_message=f"fail to run {cmd}",
+        )
+        matched_results = self._result_pattern.match(result.stdout)
+        assert matched_results
+        iops = matched_results.group("iops")
+        if iops.endswith("k"):
+            iops_value = Decimal(iops[:-1]) * 1000
+        else:
+            iops_value = Decimal(iops)
+        latency = matched_results.group("latency")
+        fio_result = FIOResult()
+        fio_result.iops = iops_value
+        fio_result.latency = Decimal(latency)
+        fio_result.iodepth = iodepth
+        fio_result.qdepth = iodepth * numjob
+        fio_result.mode = mode
+        return fio_result
+
+    def parse_test_results(
+        self, fio_results_list: List[FIOResult]
+    ) -> List[FIOResultsMessage]:
+        qdepth_list = sorted(set([x.qdepth for x in fio_results_list if x.qdepth]))
+        fio_message: List[FIOResultsMessage] = []
+        for qdepth in qdepth_list:
+            mode_iops_latecny = self._get_iops_latency_by_qdepth(
```

Comment:
It may not need the extra loop. Just construct a dict of qdepth/message, and then extra the message list from the dict. One loop should be enough.

Reply:
when construct a dict of qdepth/message we still need loop? for every qdepth, there are 4 modes' iops and latency data. then the value of dict qdepth/message need to be an object?

Reply:
The qdepth/message dict is used to merge data. One for loop is still necessary to go through all combinations.

Reply:
updated